types/src/transaction/mod.rs

```diff
@@ -538,6 +762,158 @@ impl TransactionPayload {
             payload => panic!("Expected EntryFunction(_) payload, found: {:#?}", payload),
         }
     }
+
+    pub fn has_nonce(&self) -> bool {
+        match self {
+            Self::Payload(TransactionPayloadInner::V1 {
+                extra_config:
+                    TransactionExtraConfig::V1 {
+                        replay_protection_nonce,
+                        ..
+                    },
+                ..
+            }) => replay_protection_nonce.is_some(),
+            _ => false,
+        }
+    }
+
+    pub fn executable(&self) -> TransactionExecutable {
+        match self {
+            TransactionPayload::EntryFunction(entry_function) => {
+                TransactionExecutable::EntryFunction(entry_function.clone())
+            },
+            TransactionPayload::Script(script) => TransactionExecutable::Script(script.clone()),
+            TransactionPayload::Multisig(multisig) => multisig.as_transaction_executable(),
+            TransactionPayload::Payload(TransactionPayloadInner::V1 { executable, .. }) => {
+                executable.clone()
+            },
+            TransactionPayload::ModuleBundle(_) => {
+                unimplemented!("ModuleBundle variant is deprecated")
+            },
+        }
+    }
+
+    pub fn executable_ref(&self) -> Result<TransactionExecutableRef> {
+        match self {
+            TransactionPayload::EntryFunction(entry_function) => {
+                Ok(TransactionExecutableRef::EntryFunction(entry_function))
+            },
+            TransactionPayload::Script(script) => Ok(TransactionExecutableRef::Script(script)),
+            TransactionPayload::Multisig(multisig) => Ok(multisig.as_transaction_executable_ref()),
+            TransactionPayload::Payload(TransactionPayloadInner::V1 { executable, .. }) => {
+                Ok(executable.as_ref())
+            },
+            TransactionPayload::ModuleBundle(_) => {
+                Err(format_err!("ModuleBundle variant is deprecated"))
+            },
+        }
+    }
+
+    pub fn extra_config(&self) -> TransactionExtraConfig {
+        match self {
+            TransactionPayload::Script(_)
+            | TransactionPayload::EntryFunction(_)
+            | TransactionPayload::ModuleBundle(_) => TransactionExtraConfig::V1 {
+                multisig_address: None,
+                replay_protection_nonce: None,
+            },
+            TransactionPayload::Multisig(multisig) => TransactionExtraConfig::V1 {
+                multisig_address: Some(multisig.multisig_address),
+                replay_protection_nonce: None,
+            },
+            TransactionPayload::Payload(TransactionPayloadInner::V1 { extra_config, .. }) => {
+                extra_config.clone()
+            },
+        }
+    }
+
+    // Used in transaction emitter when printing errors.
+    pub fn payload_type(&self) -> Cow<'static, str> {
```

Comment:
why this is a Cow?

Reply:
I actually used to return `String` here. Alden commented to use `Cow` instead here. 

https://github.com/aptos-labs/aptos-core/pull/16012#discussion_r1985781219